### PR TITLE
fix: reduce config fields used in branch fingerprint 

### LINKED ISSUE #18353: Reduce BranchConfig used to compute skipBranchUpdate
### Describe the proposed change(s).

We use `branchFingerprint` to determine whether an existing branch should be updated (i.e. clone + write updates + compare + commit).
We use the `branchConfig` in the calculation of the `branchFingerprint` so that if any field of the config differs, the fingerprint will be different and we will perform an update.

The issue is that we set the entire `branchConfig`, which contains many fields that are irrelevant to the branch-update process and are used for validation, scheduling, and managing PRs. When such a "config-field" changes, the fingerprint changes, and we end up running through the while update process and cloning the repo, increasing repo processing time.

We can avoid the problem by creating a new config and using it to compute `branchFingerprint` after filtering out all unnecessary fields from `branchConfig`.

It's worth noting that the current `branchConfig` contains all of the fields present in the `RenovateConfig` because we never narrow down the config when passing it to processBranch(), which increases the chances of unwanted fingerprint changes.

The solution is straightforward. Make a list of all the fields that have an impact on the branch-update process and use it when fingerprinting.
Let us call the new configuration -> `fingerprintConfig`, and then we must:
1. Narrow `BranchConfig` down to `fingerprintConfig`. We filter out all `BranchConfig` fields that aren't used in branch processing.
2. Further restrict `fingerprintConfig` to only include fields used in the branch-update process, i.e. remove config related to scheduling, validation, etc.

### PR DESCRIPTION
**Pull request by RahulGautamSingh**
## Changes
- Use `branchConfig.upgrades` to calculate fingerprint
- Filter `upgrades` to only include fields that give details related version and grouping.
## Context
- Closes : https://github.com/renovatebot/renovate/issues/18353

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [x] No documentation update is required

## How I've tested my work (please tick one)

I have verified these changes via:

- [x] Code inspection only, or
- [ ] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [ ] Both unit tests + ran on a real repository

### CODE REVIEW

**Comment by RahulGautamSingh** [comment]
Accidentally created it from different branch.

**Comment by RahulGautamSingh** [comment]
I haven't included the config-fields which effect the commitMessage assuming that if a user wants commit message to change then it will be expected in new PRs rather than ones that have already been created.

**Comment by RahulGautamSingh** [comment]
List of reason I could think of that will cause renovate-bot to make a new commit on an open PR:
1. rebase due to modified, conflicted, behind-base, user-requested
2. When using same branch to create a new PR ie. new version available while PR which updates same dep but old version is not merged.

If these are the noly 2 cases then we can just fingerprint `upgrades` field to decide whether to skipBranchUpdate or and `reuseExistingBranch` will handle the other conditions.

File: lib/workers/repository/process/fingerprint-fields.ts [hunk 1]
```diff
@@ -0,0 +1,187 @@
+export const actualBranchConfigFields = [
```
Comment (rarkins) [inline lib/workers/repository/process/fingerprint-fields.ts:1]: This is unmaintainable, we should not continue down this approach. Please stop work and go back to the issue to discuss approaches

File: lib/workers/repository/process/fingerprint-fields.ts [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+  lockedVersion?: string | null;
+  newVersion?: string;
+  newDigest?: string;
+  newName?: string;
+  updateType?: UpdateType;
+  replaceString?: string;
+  extractVersion?: string;
+}
+
+export const upgradeFingerprintFields = [
```
Comment (rarkins) [inline lib/workers/repository/process/fingerprint-fields.ts:21]:
```suggestion
type K = keyof UpgradeFingerprintConfig;

export const upgradeFingerprintFields: K[] = [
```

File: lib/workers/repository/process/fingerprint-fields.ts [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+  digestOneAndOnly?: boolean;
+  fixedVersion?: string;
+  currentVersion?: string;
+  lockedVersion?: string | null;
+  newVersion?: string;
+  newDigest?: string;
+  newName?: string;
+  updateType?: UpdateType;
+  replaceString?: string;
+  extractVersion?: string;
```
Comment (rarkins) [inline lib/workers/repository/process/fingerprint-fields.ts:18]: Are you sure `extractVersion` is used during update? I thought only during lookup

**Review by rarkins — CHANGES_REQUESTED** [review]

File: lib/workers/repository/process/fingerprint-fields.ts (hunk 3 above)
Comment (RahulGautamSingh) [reply lib/workers/repository/process/fingerprint-fields.ts:18]: Not sure. I was just not sure about the difference between extractVersion and currentVersion so included it so we can discuss/

File: lib/workers/repository/process/fingerprint-fields.ts [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+  currentVersion?: string;
+  lockedVersion?: string | null;
+  newVersion?: string;
+  newDigest?: string;
+  newName?: string;
+  updateType?: UpdateType;
+  replaceString?: string;
+}
+
+type K = keyof UpgradeFingerprintConfig;
```
Comment (rarkins) [inline lib/workers/repository/process/fingerprint-fields.ts:20]: Move these into a types.ts file

File: lib/workers/repository/process/fingerprint-fields.ts [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+
+export const upgradeFingerprintFields: K[] = [
```
Comment (rarkins) [inline lib/workers/repository/process/fingerprint-fields.ts:22]:
```suggestion
type BranchFingerprintFields = keyof UpgradeFingerprintConfig;

export const upgradeFingerprintFields: BranchFingerprintFields[] = [
```

File: lib/workers/repository/process/fingerprint-fields.ts [hunk 6]
```diff
@@ -0,0 +1,20 @@
+import type { UpgradeFingerprintConfig } from '../../types';
+
+type BranchFingerprintFields = keyof UpgradeFingerprintConfig;
+
+export const upgradeFingerprintFields: BranchFingerprintFields[] = [
+  'manager',
+  'packageFile',
+  'lockFile',
+  'currentDigestShort',
+  'datasource',
```
Comment (rarkins) [inline lib/workers/repository/process/fingerprint-fields.ts:10]:
```suggestion
  'datasource',
  'depName',
  'packageName',
  'currentValue',
  'newValue',
```

File: lib/workers/repository/process/fingerprint-fields.ts [hunk 7]
```diff
@@ -0,0 +1,20 @@
+import type { UpgradeFingerprintConfig } from '../../types';
+
+type BranchFingerprintFields = keyof UpgradeFingerprintConfig;
+
+export const upgradeFingerprintFields: BranchFingerprintFields[] = [
+  'manager',
+  'packageFile',
+  'lockFile',
+  'currentDigestShort',
```
Comment (rarkins) [inline lib/workers/repository/process/fingerprint-fields.ts:9]:
```suggestion
```

File: lib/workers/repository/process/fingerprint-fields.ts [hunk 8]
```diff
@@ -0,0 +1,20 @@
...
+  'datasource',
+  'digestOneAndOnly',
```
Comment (rarkins) [inline lib/workers/repository/process/fingerprint-fields.ts:11]:
```suggestion
```

File: lib/workers/repository/process/fingerprint-fields.ts [hunk 9]
```diff
@@ -0,0 +1,20 @@
...
+  'fixedVersion',
+  'currentVersion',
+  'lockedVersion',
+  'newVersion',
+  'newDigest',
+  'newName',
+  'updateType',
```
Comment (rarkins) [inline lib/workers/repository/process/fingerprint-fields.ts:18]:
```suggestion
```

**Review by rarkins — CHANGES_REQUESTED** [review]

File: lib/workers/repository/process/fingerprint-fields.ts [hunk 10]
```diff
@@ -0,0 +1,16 @@
...
+  'manager',
+  'packageFile',
+  'lockFile',
+  'fixedVersion',
+  'currentVersion',
+  'lockedVersion',
+  'newVersion',
+  'newDigest',
+  'newName',
+  'replaceString',
```
Comment (rarkins) [inline lib/workers/repository/process/fingerprint-fields.ts:15]:
```suggestion
  'autoReplaceStringTemplate',
  'currentDigest',
  'currentValue',
  'currentVersion',
  'datasource',
  'depName',
  'lockFile',
  'lockedVersion',
  'manager',
  'newName',
  'newDigest',
  'newValue',
  'newVersion',
  'packageFile',
  'packageName',
  'replacementName',
  'replacementVersion',
  'replaceString',
```

File: lib/workers/repository/process/write.spec.ts [hunk 11]
```diff
@@ -197,7 +198,7 @@ describe('workers/repository/process/write', () => {
         ),
       ].sort();
       const branchFingerprint = fingerprint({
-        branch,
+        ...generateBranchFingerprintConfig(branch),
```
Comment (rarkins) [inline lib/workers/repository/process/write.spec.ts:201]:
```suggestion
        branchFingerprintConfig: generateBranchFingerprintConfig(branch),
```

File: lib/workers/repository/process/write.ts [hunk 12]
```diff
@@ -8,12 +8,31 @@ import { fingerprint } from '../../../util/fingerprint';
...
 
+export function generateBranchFingerprintConfig(
+  branch: BranchConfig
+): UpgradeFingerprintConfig[] {
+  const res = branch.upgrades.map((upgrade) => {
+    const filtertedUpgrade = {} as any;
+    for (const field of upgradeFingerprintFields) {
+      filtertedUpgrade[field] = upgrade[field];
+    }
+    return filtertedUpgrade as UpgradeFingerprintConfig;
```
Comment (rarkins) [inline lib/workers/repository/process/write.ts:30]: This kind of destroys the point of types. Maybe Partial<> works better? @JamieMagee @viceice @zharinov are more experts on this than me

File: lib/workers/repository/process/write.ts [hunk 13]
```diff
@@ -129,7 +148,7 @@ export async function writeUpdates(
       ),
     ].sort();
     const branchFingerprint = fingerprint({
-      branch,
+      ...generateBranchFingerprintConfig(branch),
```
Comment (rarkins) [inline lib/workers/repository/process/write.ts:151]:
```suggestion
      branchFingerprintConfig: generateBranchFingerprintConfig(branch),
```

File: lib/workers/types.ts [hunk 14]
```diff
@@ -152,3 +152,16 @@ export interface SelectAllConfig extends RenovateConfig {
   dependencyDashboardAllPending?: boolean;
   dependencyDashboardAllRateLimited?: boolean;
 }
+
+export interface UpgradeFingerprintConfig {
```
Comment (rarkins) [inline lib/workers/types.ts:156]: This will need updating after my changes

**Review by rarkins — CHANGES_REQUESTED** [review]

File: lib/workers/repository/process/write.ts (hunk 12 above)
Comment (viceice) [reply lib/workers/repository/process/write.ts:30]: i don't think we need the any and can use the type directly, as we cop only field of this config type. 🤔

Comment (RahulGautamSingh) [reply lib/workers/repository/process/write.ts:30]: Throws error:
image

Comment (viceice) [reply lib/workers/repository/process/write.ts:30]: you need to make sure every field is also in the original upgrade config object

File: lib/workers/repository/process/write.ts [hunk 15]
```diff
@@ -8,12 +8,31 @@ import { fingerprint } from '../../../util/fingerprint';
...
 
+export function generateBranchFingerprintConfig(
+  branch: BranchConfig
+): UpgradeFingerprintConfig[] {
+  const res = branch.upgrades.map((upgrade) => {
+    const filtertedUpgrade = {} as UpgradeFingerprintConfig;
+    for (const field of upgradeFingerprintFields) {
+      filtertedUpgrade[field] = upgrade[field];
+    }
+    return filtertedUpgrade;
```
Comment (rarkins) [inline lib/workers/repository/process/write.ts:30]:
```suggestion
    const filteredUpgrade = {} as UpgradeFingerprintConfig;
    for (const field of upgradeFingerprintFields) {
      filteredUpgrade[field] = upgrade[field];
    }
    return filteredUpgrade;
```

**Review by rarkins — CHANGES_REQUESTED** [review]
Otherwise LGTM

**Review by rarkins — APPROVED** [review]

**Pull request merged**